torchsnapshot/storage_plugins/fsspec.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import asyncio
+import io
+import os
+
+import fsspec
+
+from torchsnapshot.io_types import ReadIO, StoragePlugin, WriteIO
+
+__all__ = ["FSSpecStoragePlugin"]
+
+
+class FSSpecStoragePlugin(StoragePlugin):
+    def __init__(self, root: str) -> None:
+        root_items = root.split("://")
+        if len(root_items) != 2:
+            raise ValueError("only protocol://path is supported by fsspec plugin")
+        protocol, self.root = root_items
+        if not protocol.startswith("fsspec-"):
+            raise ValueError(
+                f"Invalid protocol: {protocol}, Only fsspec-* protocols are supported"
+            )
+        self.fs = fsspec.filesystem(protocol=protocol.removeprefix("fsspec-"))
+        self._session = None
+        self._lock = asyncio.Lock()
+
+    async def _init_session(self) -> None:
+        async with self._lock:
+            if self._session is None:
+                self._session = await self.fs.set_session()
+
+    async def write(self, write_io: WriteIO) -> None:
```

Comment:
`StoragePlugin.write` is responsible for creating parent directories if there are any. For example, if `write_io.path` is `foo/bar/baz`, the method needs to ensure that both `foo` and `foo/bar` exist.

There's a caveat: not all `fsspec` plugins support `mkdir`. One example I can think of is `s3fs`. For such plugins, we don't need to create directories, and we can't call `mkdir` which will raise an exception.

Reply:
you're right, the write method should be responsible for creating parent path and not all plugins support mkdir. 

I think we can raise an exception if the plugin doesn't have mkdir method.

But it seems that the s3fs also provide the mkdir method. 

The implement is here:

https://github.com/fsspec/s3fs/blob/main/s3fs/core.py#L803

Reply:
> But it seems that the s3fs also provide the mkdir method

Thanks for the pointer. Apparently I remembered it wrong.

> I think we can raise an exception if the plugin doesn't have mkdir method.

If a plugin doesn't implement `mkdir`, most likely it's not needed and I think we can just suppress the exception.